near/src/validator_manager.rs

```diff
@@ -327,21 +367,77 @@ impl ValidatorManager {
         new_hash: CryptoHash,
     ) -> Result<(), ValidatorError> {
         let mut proposals = vec![];
+        let mut validator_tracker = HashMap::new();
         let mut hash = last_hash;
+        let prev_epoch_hash = self.get_prev_epoch_hash(epoch_hash)?;
+        // TODO: cache the hashmaps
+        let (block_index_to_validator, validator_to_num_blocks) = {
+            let epoch_index_info = self.get_index_info(epoch_hash)?;
+            let epoch_length = self.config.epoch_length;
+            let validator_assignment = self.get_validators(prev_epoch_hash)?;
+            get_epoch_block_proposer_info(
+                validator_assignment,
+                epoch_length,
+                epoch_index_info.index,
+            )
+        };
+
         loop {
             let info = self.get_index_info(hash)?;
             if info.epoch_start_hash != epoch_hash || info.prev_hash == hash {
                 break;
             }
             proposals.extend(info.proposals);
+            // safe to unwrap because block_index_to_validator is computed from indices in this epoch
+            let validator = *block_index_to_validator.get(&info.index).unwrap();
+            validator_tracker.entry(validator).and_modify(|e| *e += 1).or_insert(1);
             hash = info.prev_hash;
         }
         let mut store_update = self.store.store_update();
 
+        let mut last_epoch_proposals = self
+            .store
+            .get_ser(COL_LAST_EPOCH_PROPOSALS, prev_epoch_hash.as_ref())?
+            .unwrap_or_else(|| vec![]);
+        store_update.set_ser(COL_LAST_EPOCH_PROPOSALS, epoch_hash.as_ref(), &proposals)?;
+        proposals.append(&mut last_epoch_proposals);
+
+        let validator_kickout: HashMap<AccountId, bool> = {
+            let validator_kickout_threshold = self.config.validator_kickout_threshold;
+            let validator_assignment = self.get_validators(prev_epoch_hash)?;
+            let mut all_kicked_out = true;
```

Comment:
hm, so one idea to remove this hack is to count "num_blocks_expect" based on the maximum blocks one seat produced. This way even if everybody reduced their production due to some DDOS attack on the network or one guy left and producing blocks really infrequently - it will keep that going without additional logic.
 
Don't need to change this PR, but something to consider.

Reply:
Wouldn't this lead to gameable behavior?

Reply:
So one problem this might have is that, if the epoch length is small then the difference in the number of blocks validators should produce might cause wrong kick-outs. For example, if some validator produced 5 blocks in one epoch but another one is only expected to produce 4, then that validator will be kicked out no matter how many blocks they produce if the threshold is 0.9.